# Exploring Hacker News Posts

In this project, we will analyze a dataset of submissions to the popular technology site [Hacker News](https://news.ycombinator.com/). Hacker News is a platform where user-submitted stories ("posts") receive votes and comments.

We are specifically interested in two types of posts:
* **Ask HN**: Posts submitted to ask the community a specific question.
* **Show HN**: Posts submitted to show the community a project or something interesting.

### Our Goals
1. Determine if `Ask HN` or `Show HN` receive more comments on average.
2. Determine if posts created at `a certain` time receive more comments on average.

### Dataset Description
The dataset consists of approximately 20,000 rows, each representing a post with attributes such as the title, number of points, number of comments, and submission timestamp.

## 1. Loading the Dataset
To begin, we need to import and read the hacker_news.csv file into Python as a list of lists.

In [1]:
from csv import reader 
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)

headers = hn[0]  # assigning the header row
hn = hn[1:]  # removing the header row


hn[:5]  #checking

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

In [2]:
headers  # checking

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

## 2. Separating Ask HN and Show HN Posts

Now that the data is prepared, we will filter the submissions into three distinct lists. Using the `startswith()` string method and the `lower()` method ensures that we capture all relevant posts regardless of their capitalization.

* `ask_posts`: Contains posts where users ask questions.
* `show_posts`: Contains posts where users showcase their work.
* `other_posts`: Contains all other submissions.

In [3]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1].lower()
    if title.startswith("ask hn"):
        ask_posts.append(row)
    elif title.startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)

# Verifying the results by checking the lengths of the lists
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


## 3. Calculating the Average Number of Comments

Now that we have separated the "Ask" and "Show" posts, we will calculate the average number of comments for each category. This will help us determine which type of post generally generates more engagement from the Hacker News community.

In [4]:
# Calculate average comments for Ask HN posts
total_ask_comments = 0

for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments

avg_ask_comments = total_ask_comments / len(ask_posts)
print("Average comments on Ask HN posts:", avg_ask_comments)

# Calculate average comments for Show HN posts
total_show_comments = 0

for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments

avg_show_comments = total_show_comments / len(show_posts)
print("Average comments on Show HN posts:", avg_show_comments)

Average comments on Ask HN posts: 14.038417431192661
Average comments on Show HN posts: 10.31669535283993


## 4. Comparison of Findings

Based on the analysis:
* **Ask HN** posts receive approximately **14.04** comments on average.
* **Show HN** posts receive approximately **10.32** comments on average.

Our analysis shows that **Ask HN** posts receive more comments on average than **Show HN** posts. Since our dataset focused on posts that received at least one comment, we can conclude that the community is more likely to engage with questions than with project showcases.

## 5. Analyzing Engagement by Hour of Submission

Since `Ask HN` posts receive more engagement on average, we will focus our remaining analysis on this category. We aim to determine if there is a specific time of day when submitting a post results in a higher number of comments.

We will use the `datetime` module to isolate the hour from the `created_at` column and then aggregate the number of posts and comments for each hour.

In [5]:
import datetime as dt

# Create a list containing only the creation time and number of comments
result_list = []

for row in ask_posts:
    created_at = row[6]
    n_comments = int(row[4])
    result_list.append([created_at, n_comments])

counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for row in result_list:
    date_str = row[0]
    comment = row[1]
    
    # Parse the date string and extract the hour
    date_dt = dt.datetime.strptime(date_str, date_format)
    hour = date_dt.strftime("%H")
    
    # Update frequency tables
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comment
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comment

print("Number of posts created per hour:", counts_by_hour)
print("\nTotal comments received per hour:", comments_by_hour)

Number of posts created per hour: {'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}

Total comments received per hour: {'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


---
With the frequency tables created, we now have the total number of posts and comments for each hour. In the next step, we will use these dictionaries to calculate the average number of comments per post for each hour of the day.

## 6. Calculating the Average Comments per Hour

With the frequency tables for the number of posts and total comments per hour, we can now calculate the average number of comments for each hour. We will store these results in a list of lists where each inner list contains the hour and its corresponding average comment count.

In [6]:
avg_by_hour = []

for hour in comments_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour] / counts_by_hour[hour]])

# Display the results
print(avg_by_hour)

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


### Analyzing the Raw Averages

The list above displays the average number of comments per post for each hour of the day. While this data is accurate, it is currently difficult to read and is not sorted in a way that allows us to easily identify the hours with the highest engagement. In the next step, we will sort and format these results for better clarity.

## 7. Sorting and Identifying the Top Hours for Engagement

To easily identify the hours with the highest average number of comments, we will swap the positions of the hour and the average value in our list. This allows us to use Python's `sorted()` function to sort by the average number of comments in descending order. Finally, we will format the output to make it professional and readable.

In [7]:
# Swap the columns to sort by the average number of comments
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])

# Sort the list in descending order
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

print("Top 5 Hours for Ask Posts Comments")

# Format and print the top 5 results
for avg, hr in sorted_swap[:5]:
    # Parse the hour string and format it
    hr_dt = dt.datetime.strptime(hr, "%H")
    hr_str = hr_dt.strftime("%H:%M")
    
    print(f"{hr_str}: {avg:.2f} average comments per post")

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


## 8. Conclusion and Findings

Based on our analysis of the top 5 hours for comments on `Ask HN` posts:
1. **15:00**: 38.59 average comments per post
2. **02:00**: 23.81 average comments per post
3. **20:00**: 21.53 average comments per post
4. **16:00**: 16.80 average comments per post
5. **21:00**: 16.01 average comments per post

The hour that receives the most comments on average is **15:00 (3:00 PM EST)**. 

## 9. Recommendations
To maximize the chance of receiving comments:
* format the post as an "Ask HN",
* consider posting at **15:00 EST**.

  According to the dataset documentation, the times are recorded in Eastern Time in the US. For my local timezone (Karakol, Kyrgyzstan), 15:00 EST corresponds to **02:00 AM** the following day. 

In summary, the community is most active in terms of discussion during the mid-afternoon hours (Eastern Time), with a secondary peak during the late evening.